setup/vars.tf

```diff
@@ -29,6 +40,20 @@ variable "project_id" {
   }
 }
 
+# This variable is used to override default naming behavior.  By default, names
+# are scoped by terraform.workspace if not "default", otherwise it will be set
+# to "prod" if is_production is true or "nonprod" if false
+variable "environment" {
+  type        = string
+  default     = "default"
+  description = "The name of the environment"
+
+  validation {
```

Comment:
I really wish that you could validate against another variable bc this `is_production` and `environment` variable dependency is very confusing, as is the special meaning of "default".  Also it is confusing that the presence of the key `is_production` does not have the same behavior as `is_production = false`. From SO it doesn't seem possible without more hacks though: https://stackoverflow.com/questions/63629916/terraform-13-validate-variable-base-on-the-value-of-another

Any thoughts on best approach here?

edit: reading the code more naming this `custom_env_override` may help?

edit edit: is it ever even possible to have a custom env name? is_production defaults to false, so it will always be set.

Reply:
also putting upfront that it is only applicable if is_production is not set in the comment

Reply:
Yeah, I kinda went back and forth on a few different options here.

As with the other Terraform projects, `environment` is more for namespacing and tagging than anything else.  It doesn't have a material effect on resource creation and can be arbitrarily named (within certain guidelines).  In most deployment scenarios, including the one we will be using for DHP, there will only be two instances of this project: "prod" and "nonprod".  I don't want to force that restriction on anyone else who might use this in the future once it gets pulled into bloom-infra, so the environment variable serves as an override.  To your point, perhaps a name like "env_override" or something like that would make this a bit more clear, but this is basically just a standard "set it if you want something other than the default" variable.

One way that default value is determined is to try to use the value of `terraform.workspace` if set.  If a custom workspace has not been set, this value defaults to "default".  By keeping the default value for `environment` set to "default", it then becomes a very simple check to determine whether both the `environment` var and `terraform.workspace` are set to their default values (line 33).  If it still hasn't been set, pick one based on the value of `is_production`.

It is important to note that `is_production` is independent from the environment name and is intended to toggle certain behavior.  I'm generally not a big fan of magic toggles that control undefined settings, but considering that there are only two known valid states for the resources created by this project, that the number of settings are currently unknown but could potentially get rather large, and that it's assumed that our target users would not have the experience to be able to pick the right settings themselves, I just threw it in.  I figured if nothing else I could just do a find/replace later with more specific options if needed.

It's important to note that there isn't a reasonable deployment scenario that doesn't involve the use of workspaces.    If there is more than one instance, the state files need to be kept separate somehow, so it's reasonable to assume that most implementers will just set their workspace and ignore the `environment` var.  There are still scenarios where you might want to override that, but it's really only edge cases where you would need to have an additional fallback to that.